Lock down the access to the member variables in ADContext.

bgogul opened the pull request:
This PR locks down the internals of ADContext via appropriate API interface. This would be important to hide the details of ADContext when we move this out to a separate file. 

(This PR is a part of https://bugs.swift.org/browse/TF-993)

bgogul committed:
Lock down the access to the member variables in ADContext.

bgogul commented:
@swift-ci please test tensorflow

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -838,36 +838,67 @@ class ADContext {
+  /// Returns nullptr if none are available for processing.
+  DifferentiableFunctionInst* pickAndRemoveDifferentiableFunctionInst() {
+    if (differentiableFunctionInsts.empty()) return nullptr;
+    auto *dfi = differentiableFunctionInsts.back();
```
I believe you can replace these three lines with `differentiableFunctionInsts.pop_back_val()`, which pops and returns the back value.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -838,36 +838,67 @@ class ADContext {
+
+  bool
+  isDifferentiableFunctionInstProcessed(DifferentiableFunctionInst *dfi) const {
+    return processedDifferentiableFunctionInsts.count(dfi) > 0;
```
`count` is the name for `contains` in LLVM APIs. Here, it effectively returns a `bool`, so it's idiomatic to use its result as a `bool` directly.

```suggestion
    return processedDifferentiableFunctionInsts.count(dfi);
```

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -838,36 +838,67 @@ class ADContext {
+  }
+
+  /// Updates the index for `dfi` in this context.
+  unsigned updateResultIndex(DifferentiableFunctionInst* dfi, unsigned index) {
```
Please make this a `void` method, since the result is never used.

I think "set" is more apt than "update" here. Result indices are really meant to be stored in `DifferentiableFunctionInst`, so they shouldn't be updated with a new different value.
```cpp
  /// Set the result index for `dfi`.
  void setResultIndex(DifferentiableFunctionInst* dfi, unsigned index) {
```

bgogul replied:
Oops! Fixed now.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -838,36 +838,67 @@ class ADContext {
-    return processedDifferentiableFunctionInsts;
+  /// Pick and remove a `DifferentiableFunctionInst` for processing.
+  /// Returns nullptr if none are available for processing.
+  DifferentiableFunctionInst* pickAndRemoveDifferentiableFunctionInst() {
```
How about this more descriptive name?
```suggestion
  DifferentiableFunctionInst* popDifferentiableFunctionInstFromWorklist() {
```
"pop" is a shorter term-of-art than "pick and remove". "from worklist" clarifies where the `DifferentiableFunctionInst` came from.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -838,36 +838,67 @@ class ADContext {
 
-  SmallVector<SILFunction *, 32> &getGeneratedFunctions() {
-    return generatedFunctions;
+  void addGeneratedFunction(SILFunction* function) {
```
I think "record" is slightly more apt.
```suggestion
  void recordGeneratedFunction(SILFunction* function) {
```

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -838,36 +838,67 @@ class ADContext {
 
-  SmallVector<SILValue, 32> &getGeneratedFunctionReferences() {
-    return generatedFunctionReferences;
+  void addGeneratedFunctionReference(SILValue functionRef) {
```
```suggestion
  void recordGeneratedFunctionReference(SILValue functionRef) {
```

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -838,36 +838,67 @@ class ADContext {
-  getDifferentiableFunctionInsts() {
-    return differentiableFunctionInsts;
+  /// Returns true if there are `DifferentiableFunctionInst` to process.
+  bool hasDifferentiableFunctionInsts() const {
```
How about:
```suggestion
  bool isDifferentiableFunctionInstWorklistEmpty() const {
```
Mentioning "worklist" clarifies where the `DifferentiableFunctionInst` came from.
With this new name, the method body should be the inverse: `return differentiableFunctionInsts.empty();`.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -838,36 +838,67 @@ class ADContext {
+  /// Returns the index for `dfi` if found in this context. Otherwise,
+  /// sets the index to zero and returns it.
+  unsigned getResultIndex(DifferentiableFunctionInst* dfi) {
+    return resultIndices[dfi];
```
Could you please add `assert(resultIndices.count(dfi) && "No result index set")`?

Please also use `resultIndices.lookup(dfi)` instead of `resultIndices[dfi]`.
`lookup` does not emplace the default constructed value in the map, while the subscript operator does.

dan-zheng reviewed (commented):
Some minor comments!

bgogul committed:
Address review comments regarding names and other concerns.

bgogul commented:
Addressed the review comments. PTAL.

bgogul commented:
@swift-ci please test tensorflow

dan-zheng committed:
Minor edits.

dan-zheng commented:
> Addressed the review comments. PTAL.

To save our time, I added a NFC commit: https://github.com/apple/swift/pull/28567/commits/af9f0178444928a936739d5726e0828615ea03b2. Hope that's fine!

bgogul commented:
> > Addressed the review comments. PTAL.
> 
> To save our time, I added a NFC commit: [af9f017](https://github.com/apple/swift/commit/af9f0178444928a936739d5726e0828615ea03b2). Hope that's fine!

Of course, thanks!

bgogul committed:
Fix the CI failure.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -838,36 +840,70 @@ class ADContext {
-                       32> &
-  getInvokers() {
+  /// Adds the given `differentiable_function` instruction to the worklist.
+  void addDifferentiableFunctionInst(DifferentiableFunctionInst* dfi) {
```
minor, for consistency:
```suggestion
  void addDifferentiableFunctionInst(DifferentiableFunctionInst *dfi) {
```
Could you please update the rest of the added helper functions taking pointer-typed arguments?

bgogul replied:
Hmm...I thought I did a `clang-format`. Anyway, I will fix them.

dan-zheng replied:
By default, `clang-format` seems to leave `*` and `&` positions untouched. There are configuration fields like `PointerAlignment` though: https://clang.llvm.org/docs/ClangFormatStyleOptions.html

bgogul committed:
Fix pointer style.

bgogul commented:
@swift-ci please test tensorflow

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -838,36 +840,70 @@ class ADContext {
-                       32> &
-  getInvokers() {
+  /// Adds the given `differentiable_function` instruction to the worklist.
+  void addDifferentiableFunctionInst(DifferentiableFunctionInst* dfi) {
```
```suggestion
  void addDifferentiableFunctionInstToWorklist(DifferentiableFunctionInst* dfi) {
```
Sorry for the late comment, feel free to address in a follow-up!

bgogul replied:
OK. will address in a follow up.
